docs/architecture/adr-007.md

```diff
@@ -0,0 +1,96 @@
+
+## Decision
+
+We define 2 specifics events that smart contract can use to trigger specific action after contract execution.
+
+```
+# Trigger an action after contract execution which only involve the sender account
+__CronosCall(string functionName, address sender,  byte[]args)
+
+# Trigger an action after contract execution that require moving asset from the smart contract "account"
+__CronosCallAfterPayment(string functionName, address payer, uint amount, byte[]args)
+
+```
+
+- **functionName** corresponds to the equivalent hardcoded eventName such as `CronosSendToIBC` for the event `__CronosSendToIbc` or `CronosSendToChain` for the event `__CronosSendToChain`
+- **payer** is the address which make the payment, usually equal to `msg.sender`
+- **amount** corresponds to the `amount` that the payer has paid (either by burning or sending asset to the SC)
+- **args** corresponds to the argument in order of the event removing the `sender` and `amount` from the parameters encoded using the abi event pack
+
+
+
+An example of smart contract implementing those function
+
+```
+pragma solidity ^0.6.8;
+
+import "ds-token/token.sol";
+
+Contract Demo is DSToken {
+
+ __CronosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)
+ 
+ // unsafe_burn burn tokens without user's approval and authentication, used internally
+ function unsafe_burn(address addr, uint amount) internal {
+     // Deduct user's balance without approval
+     require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
+     balanceOf[addr] = sub(balanceOf[addr], amount);
+     totalSupply = sub(totalSupply, amount);
+     emit Burn(addr, amount);
+ }
+ 
+ // send an "amount" of the contract token to recipient through IBC
+ // args should contains the encoded value recipient or extra value if needed in the future
+ function send_to_ibc(uint amount, byte[] args) public {
+     unsafe_burn(msg.sender, amount);
+     emit __CronosCallAfterPayment("CronosSendToIbc", msg.sender, amount, args);
+ }
+
+```
+
+In this example byte[] args should contains the encoded recipient address from `event __CronosSendToIbc(address sender, string recipient, uint256 amount)` but in the future, if we needs to add extra parameters such as `fee_amount` or `fee_recipient` we will be able to use the same method in the smart contract, but encode those extra parameters in args. 
+
+On the chain side, the events parameters are decoded and uses to call a specific actions. 
+
+List of possible action (as of today)
+
+- Actions : `CronosCancelSendToChain`
+- Actions after Payment: `CronosSendToAccount`, `CronosSendToIbc`, `CronosSendCroToIbc`, `CronosSendToChain`
+
+ 
+With this generic event format, it is possible to add and remove the parameters for those action in the future or even add new actions while "upgrading" the chain. Smart contract won't be force to migrate to a new contract to benefit those functionnalities.
```

Comment:
I am trying to understand how the generic can help support future event format without the need of upgrading the contract.

Contract should always validate the arguments being passed before they emit the event, right?

Reply:
no only a subset of the arguments needs to be validated. If we consider that the action is done from msg.sender account (after receiving or not cosmos asset). then most likely those fields will be specify by the users and won't need validation.

Reply:
On a second read, I think this ADR is not intended to avoid the need of contract upgrade when event signatures is changed on upgrade, am I right?

It's just providing a generic way to emit event?

Reply:
this ADR suggests a more flexible to define events for evm-hooks, and one of the issue indeed is that we cannot change event signature if we perform an upgrade and some extra parameters needs to be specified by users

Reply:
> If we consider that the action is done from msg.sender account (after receiving or not cosmos asset). then most likely those fields won't need validation.

This is true for some "chain utility" contract such as the [CroBridge](https://github.com/crypto-org-chain/cronos/blob/main/integration_tests/contracts/contracts/CroBridge.sol). 

But for most other contracts with their own functionality, I think they should safeguard the emitted events and cannot trust user inputs. And the generic event alone could not be sufficient for them. They still need to upgrade their contract to emit the new event upon chain upgrade.

Reply:
yeah this is not possible to define a single generic event that can be use for every case, and sometime upgrading the smart contract is the best way to do it. 

This ADR does not intent to solve all solution but propose a better standard at least for the token transfer functionality which involve amount deduction (burn or payment) from msg.sender and then perform an action (on the cosmos transferred amount from contract address to msg.sender address)

The action is defined by the user so contract only needs to validate the payment (amount + sender address) and user choose what action to do after

Reply:
On this angle I love the idea of generic event.

The generic event also have the added benefit of making Cosmos EVM hook from smart contract more readable --- If you see the `__CronosCall(...)` you know it is triggering an action on Cosmos layer, instead of seeing the `__CronosSendToIbc(...)` and then the reader needs to dig into different areas for what's impling. Especially in the future when there are more events need to be added.

Extending from this perspective, I am thinking if we want to define the event name to reveal this intention more explicitly, say `__InternalCall()`, `__CosmosCall()` or `__CallCosmos()` etc.

Reply:
sounds good, changed the name to `_CosmosCall`